contents/java/JUnit.md

```diff
@@ -0,0 +1,226 @@
+  public static String repeat(String reference, int copies) {
+    String result = "";
+    for (int i = 0; i < copies; i++) { // Strictly less than copies!
+      result += reference + "\n";
+    }
+    return result;
+  }
+}
+```
+
+<img src="{{baseUrl}}/contents/java/assertPass.PNG" />
+Running the same unit test again, we now have a passing test and we can be <popover effect="fade" content="We've only done one simple test. While this test works, we cannot be assured that all other cases would work too. How will our code behave if a negative number is passed in?">somewhat assured</popover> that our repeat function works as intended.
+
+<p></p> 
+
+## Benefits of JUnit
+
+_____
+
+Here are some key reasons why developers should consider using JUnit:
+
+### Benefit 1: JUnit is well-integrated with development tools
+
+JUnit is supported by many popular IDEs for developing in Java:
+* IntelliJ IDEA
+* Eclipse
+* NetBeans
+* Visual Studio Code
+
+This makes it easy for developers to write and run tests.
+
+### Benefit 2: JUnit is well-established
+
+As compared to other testing tools for Java, JUnit
+* has a longer history
+* has a larger community (As of 2019, JUnit has more questions on StackOverflow as compared to other tools)
+* is more popular among developers (according to [Synk's 2018 report](https://snyk.io/blog/jvm-ecosystem-report-2018-tools/))
+
+This means that it is easier to find answers or get help with JUnit.
+
+<p></p> 
+
+## JUnit Features
+_____
+
+In addition to the basic example in the introduction, JUnit also has more powerful features that further simplifies the process of writing complex tests.
+
+### Assertions API
+In addition to the `assertEquals` method illustrated in the introduction, JUnit allows us to check for many other things. These include:
+* `assertTrue`/`assertFalse`: checks whether statements return true/false as expected.
+* `assertNull`/`assertNotNull`: checks whether something is null or not, without causing a NullPointerException.
+* `assertArrayEquals`: yes, it loops through an entire array for us!
+* `assertThrows`/`assertDoesNotThrow`: checks that a piece of code throws the correct type of exception. In other words, JUnit even checks that the application can handle errors (or not)!
+* `assertTimeout`: ensures that a piece of code runs within time constraints - we don't want to keep the user waiting!
+
+These powerful expressions means that we can write more expressive tests.
+Instead of
+```java
+try {
+  int result = divide(100, 0); // Divides 100 by 0
+} catch (ArithmeticException ae) {
+  // This is the expected behaviour.
+  return;
+}
+// If code is here, then the ArithmeticException wasn't thrown.
+throw new AssertionError("Dividing 100 by 0 should throw an exception!");
+```
+
+We can now write
+```java
+assertThrows(ArithmeticException.class, () -> divide(100, 0));
+```
+
+This expression is a lot easier to understand, and as an added bonus, the code is also much shorter to write!
```

Comment:
This segment seems to be written as if you are explaining why JUnit5 is better than JUnit4, which does not match the target reader.